tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AWS/Kinesis/ContextPropagation.cs

```diff
@@ -0,0 +1,96 @@
+// <copyright file="ContextPropagation.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.Serialization.Formatters.Binary;
+using System.Text;
+using Datadog.Trace.Logging;
+using Datadog.Trace.Propagators;
+using Datadog.Trace.Vendors.Newtonsoft.Json;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.AWS.Kinesis
+{
+    internal static class ContextPropagation
+    {
+        private const string KinesisKey = "_datadog";
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(ContextPropagation));
+
+        public static void InjectTraceIntoRecords<TRecordsRequest>(IPutRecordsRequest request, SpanContext context)
+        {
+            // request.Records is not null and has at least one element
+            if (request.Records is not { Count: > 0 })
+            {
+                return;
+            }
+
+            var record = request.Records[0];
+            InjectTraceIntoData<TRecordsRequest>(record, context);
+        }
+
+        public static void InjectTraceIntoData<TRecordRequest>(IContainsData record, SpanContext context)
+        {
+            if (record.Data is null)
+            {
+                return;
+            }
+
+            Inject<TRecordRequest>(record, context);
+        }
+
+        public static Dictionary<string, object> MemoryStreamToDictionary(MemoryStream stream)
+        {
+            // Convert the MemoryStream to a string
+            var reader = new StreamReader(stream);
+
+            // Deserialize the JSON string into a Dictionary<string, object>
+            var dataDict = JsonConvert.DeserializeObject<Dictionary<string, object>>(reader.ReadToEnd());
```

Comment:
I see you removed the intermediate `byte[]` like we discussed, thanks!

We can go even further and also remove the `string` created by `reader.ReadToEnd()`. The `JsonConvert` type is designed as a helper to make it easy to serialize/deserialize to/from strings, but you can use `JsonSerializer` for more advanced scenarios:

```csharph
// you already wrapped the MemoryStream inside a SteamReader,
// but we need a JsonReader for JsonSerializer.Deserialize<T>()
var jsonReader = new JsonTextReader(reader);

var serializer = new JsonSerializer();
return serializer.Deserialize<Dictionary<string, object>>(jsonReader);
```

See also the section "Optimize Memory Usage" here: https://www.newtonsoft.com/json/help/html/Performance.htm

Reply:
I've updated it, please let me know what you think of the current solution!